### PR TITLE
Push message update

### LINKED ISSUE #9: Message collapsing
Need to do this once we have #3 and #6 are sorted out.

### PR DESCRIPTION
**Pull request by martinthomson**
I've specified a Topic header field that we can use as a "collapse key".

Closes #9.

### CODE REVIEW

**Comment by bbangert** [comment]
So, this changes message updating from an O(1) operation where we have an existing index to the message ID, to an O(N) operation where we have to do a full row-scan for the users notifications to find a matching topic. Adding another index for the topic is similarly a non-starter at scale.

I don't see what the problem was with the previous version that couldn't be fixed without requiring such expensive operations that this topic-based approach involves.

**Comment by bbangert** [comment]
After some more discussion with @martinthomson, I believe I can store the data in a way that provides direct index by topic as well. So I'm withdrawing my prior comment.

**Comment by bbangert** [comment]
Regarding max topics, I don't plan on limiting the max topics. I'm using a sharded db with a compound hash+range key for indexing, and plan on storing the topic in the range key for direct access to delete it for later messages with a topic field. There is the side-effect that the UA will get messages out-of-order as a result.

ie. all messages with no topic will be delivered in order, then messages with a topic will be delivered in the order of their topic names.

**Comment by beverloo** [comment]
We currently impose a limit of four _topics_ (née collapse keys) +@costinm

I do like the change. GCM has a separate feature called Topics, i.e. subscriptions to a category, which may cause some confusion. Thinking about the name without taking this into account, I still think it makes sense in this context.

https://developers.google.com/cloud-messaging/topic-messaging

**Comment by bbangert** [comment]
@beverloo yup, the common use case of sending "You have N messages in your mailbox", makes sense to just tell a developer to add the topic "Inbox", and the user will always get the last one sent in.

I am curious about how systems having topic limits plan on communicating this to the app-server. Is there some minimum the spec should mandate just so that an app developer isn't surprised later when their app needing 5 topics doesn't work right on a browser that has topic limits?

**Comment by martinthomson** [comment]
Would 429 work for systems that run out of topics (or collapse keys)?  That seems right to me.

**Comment by brianraymor** [comment]
429 seems to specific to be used for this scenario:

_The 429 status code indicates that the user has sent too many requests in a given amount of time ("rate limiting")._

Could be a 500?

**Comment by brianraymor** [comment]
Is there a length limitation on the Topic? This could also be implementation-specific and signaled by:

431 Request Header Fields Too Large (RFC6585)

**Comment by martinthomson** [comment]
> Is there a length limitation on the Topic?

I think that @bbangert and I concluded that we could accept any size of topic.  Up to the limits on header field sizes, that is, which would trigger 431.

> 429 seems too specific...

Maybe, but it is the best match I have.  And I think that 4xx is better than 5xx.

**Comment by brianraymor** [comment]
If you continue with 429, how does an application server discern between 429 Rate Limiting and 429 Too Many Topics? By the presence of Retry-After?

**Comment by martinthomson** [comment]
Retry-After might be present in all cases.  However, you raise a point that the definition of 429 leaves open:  Is this too many requests globally?  By the same client?  To the same resource?  To the same resource by the same client?

Basically, the "scope" of applicability for 429 is hard - if not impossible - to determine.  The scope is defined in some server specific terms.  The question I'd ask is whether we want to provide some webpush-specific mechanism for identifying the scope of applicability for the rejection.

**Comment by bbangert** [comment]
What about using a Scope header field in the response? In this case it will be 'Topic', maybe for too many messages for a subscription it could be 'Subscription', etc.

**Comment by martinthomson** [comment]
All these things are possible.  The question is how important it is.

Also, if we wanted to address the scope problem, I would want to address it more generally and then maybe add any push-specific extensions as necessary.  This is a problem with 429 that - if it is worth fixing here - is probably worth fixing generally.

File: draft-ietf-webpush-protocol.xml [hunk 1]
```diff
@@ -556,6 +556,95 @@ TTL = 1*DIGIT
...
+        </t>
+        <t>
+          This mechanism doesn't allow for stored push messages to be deleted.
+          An application server can send a DELETE request toward a push message
+          URL to remove (or acknowledge) an undelivered message.  Note that a
+          push message with a topic that has a zero or absent time-to-live
+          causes any stored message with the same topic to be removed, though a
+          push message with zero time-to-live could be forwarded to the user
+          agent.
+        </t>
```
Comment (hruellan) [inline draft-ietf-webpush-protocol.xml:697]: s/though a push message/though this push message/
s/could be forwarded/could still be forwarded/
Trying to make the sentence easier to grasp.

File: draft-ietf-webpush-protocol.xml [hunk 2]
```diff
@@ -556,6 +556,95 @@ TTL = 1*DIGIT
...
+ZuHSZPKa2b1jtOKLGpWrcrn8cNqt0iVQyroF
+]]></artwork>
+        </figure>
+        <figure>
+          <preamble>
+            If the push service identifies an outstanding push message with a
+            topic of "upd", the contents of that message are replaced.  The push
+            service then responds with a 201 status code, identifying the push
+            message resource that was created in a Location header field.
+          </preamble>
```
Comment (hruellan) [inline draft-ietf-webpush-protocol.xml:726]: s/The push service then responds/In all cases, the push service then responds/
To make clear that even if an outstanding push message is replaced, the push service still responds with the 201 status code.

**Comment by hruellan** [comment]
Maybe this was answered elsewhere, but why not just reusing the push message URI?
To replace a previous message, the application would create a new message and include the previous message URI in the creation request, inside a "Replace" header. Otherwise, it would work in the same way as this proposal.
This would be easy to handle on the Push Server side.
On the application server side, this would mean storing the previous message URI. But the application server needs to store the topic of the previous message anyway.

**Comment by martinthomson** [comment]
@hruellan, the idea here was that remembering the push message URI was a burden we didn't want to impose on applications.  Those that aren't using acknowledgments will be able to use this version without holding any special state.  Thanks for the grammatical review, I'll fix those up.

**Comment by hruellan** [comment]
@martinthomson, an application wanting to be able to update a message will need to keep the topic of this message, so it will need to keep this state. I don't think keeping a topic or the push message URI is a large difference. However, it's true that simple applications could use only 1 topic, which would be predefined (or a few predefined topics).

**Comment by martinthomson** [comment]
@hruellan, an application can hard-code a topic or small set of topics. They can also derive a topic from context. A URL needs to be first learned, then memorized.

**Comment by martinthomson** [comment]
@costinm, do you have any opinions on this change?

**Comment by martinthomson** [comment]
Updated based on feedback from @costinm.  Note that I haven't changed the name.  Not sure that we have consensus on that yet.

**Comment by RichardMaher** [comment]
Absolute madness! How can you be so blind, nay recalcitrant, to what Google Cloud Message and Microsoft Azure Notification Hubs are doing outside your ivory tower of Prague right now?

Mozilla's hubris in persisting with its Autopush vanity is at best aspirational and at worst just an excuse to for a whole lot of contractors/consultants to continue to plunder cash that should be invested in the flagship Firefox.

The sad truth is the abused often go one to become abusers and Mozilla abusing it's monopoly Push Service Provider position in the Firefox market turns my stomach. Tell us again about your "pledge"; Creedo? 

When with W3C peel back the scab on what is going on here?

https://github.com/webpush-wg/webpush-protocol/issues/87

***TOPICS are NOT collapse keys!***

**Pull request merged**